packages/dds/tree/src/feature-libraries/flex-map-tree/mapTreeNode.ts

```diff
@@ -0,0 +1,744 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { assert } from "@fluidframework/core-utils/internal";
+import {
+	AnchorNode,
+	EmptyKey,
+	FieldKey,
+	MapTree,
+	TreeNodeSchemaIdentifier,
+	TreeValue,
+	Value,
+} from "../../core/index.js";
+import { brand, fail, mapIterable } from "../../util/index.js";
+import {
+	FlexTreeContext,
+	FlexTreeEntityKind,
+	FlexTreeField,
+	FlexTreeFieldNode,
+	FlexTreeLeafNode,
+	FlexTreeMapNode,
+	FlexTreeNode,
+	FlexTreeNodeEvents,
+	FlexTreeOptionalField,
+	FlexTreeRequiredField,
+	FlexTreeSequenceField,
+	FlexTreeTypedField,
+	FlexTreeTypedNode,
+	FlexTreeTypedNodeUnion,
+	FlexTreeUnboxField,
+	FlexTreeUnboxNodeUnion,
+	FlexibleFieldContent,
+	FlexibleNodeSubSequence,
+	TreeStatus,
+	flexTreeMarker,
+	indexForAt,
+} from "../flex-tree/index.js";
+import {
+	FlexAllowedTypes,
+	FlexFieldNodeSchema,
+	FlexFieldSchema,
+	FlexMapNodeSchema,
+	FlexTreeNodeSchema,
+	LeafNodeSchema,
+	isLazy,
+	schemaIsFieldNode,
+	schemaIsLeaf,
+	schemaIsMap,
+	schemaIsObjectNode,
+} from "../typed-schema/index.js";
+import { FlexImplicitAllowedTypes, normalizeAllowedTypes } from "../schemaBuilderBase.js";
+import { FlexFieldKind } from "../modular-schema/index.js";
+import { FieldKinds, SequenceFieldEditBuilder } from "../default-schema/index.js";
+import { createEmitter } from "../../events/index.js";
+
+// #region Nodes
+
+/** A node's parent field and its index in that field */
+interface LocationInField {
+	readonly parent: MapTreeField<FlexAllowedTypes>;
+	readonly index: number;
+}
+
+/**
+ * A readonly implementation of {@link FlexTreeNode} which wraps a {@link MapTree}.
+ * @remarks Any methods that would mutate the node will fail,
+ * as will the querying of data specific to the {@link LazyTreeNode} implementation (e.g. {@link FlexTreeNode.context}).
+ * MapTreeNodes are unconditionally cached -
+ * when retrieved via {@link getOrCreateMapTreeNode}, the same {@link MapTree} object will always produce the same `MapTreeNode` object.
+ */
+export class MapTreeNode<TSchema extends FlexTreeNodeSchema> implements FlexTreeNode {
+	public readonly [flexTreeMarker] = FlexTreeEntityKind.Node as const;
+
+	private readonly events = createEmitter<FlexTreeNodeEvents>();
+
+	/**
+	 * Create a new MapTreeNode.
+	 * @param schema - the schema of this node
+	 * @param mapTree - the {@link MapTree} containing the data that this node wraps
+	 * @param location - the parentage of this node, if it is being created underneath an existing node and field, or undefined if not
+	 * @remarks This class (and its subclasses) should not be directly constructed outside of this module.
```

Comment:
Would it make sense to introduce interfaces for these and not export the classes at all?

Reply:
Yes, definitely, it would make sense. If it were a public type it'd be a requirement. But since it is internal, I just didn't want to add all that boilerplate. I'm not even sure if the interfaces could go in a different file or not. I could try it though. Do you think we should?

Reply:
I made a MapTreeNode interface which exposes the mapTree field (other code is going to need it). Production code will use the interface. The unit tests reach into the folder and use the stronger types.